File: packages/core/src/proofs.ts
/**
 * Proof-set verification (spec §3, §5): proofs are sorted ascending by kid,
 * unique, and must equal exactly the required signer set. Each proof signs
 * the 32 raw bytes of D(tag, body).
 */

import { b64uDecode } from "./base64url.ts";
import { digest, digestBytes, type Tag } from "./digest.ts";
import { ed25519Verify as edVerify } from "./ed25519.ts";
import type { JsonObject, JsonValue } from "./json.ts";
import { ed25519Sign as edSign } from "./ed25519.ts";

export interface SignedEnvelope {
  body: JsonObject;
  proofs: { kid: string; signature: string }[];
}

/**
 * Check the proof set equals expectedKids (distinct, compared as a set after
 * dedup of expected values), then verify every signature.
 * Returns null on success, or "PROOF_SET_INVALID" / "SIGNATURE_INVALID".
 */
export function verifyProofSet(
  tag: Tag,
  env: SignedEnvelope,
  expectedKids: string[],
  keyBytes: (kid: string) => Uint8Array | null,
  verifyFn: (pk: Uint8Array, digest: Uint8Array, sig: Uint8Array) => boolean = edVerify,
): string | null {
  const expected = [...new Set(expectedKids)].sort();
  const actual = env.proofs.map((p) => p.kid);
  const actualSorted = [...actual].sort();
  if (actual.length !== actualSorted.length || actual.some((k, i) => k !== actualSorted[i]))
    return "PROOF_SET_INVALID";
  if (new Set(actual).size !== actual.length) return "PROOF_SET_INVALID";
  if (actual.length !== expected.length || actual.some((k, i) => k !== expected[i]))
    return "PROOF_SET_INVALID";
  const digest = digestBytes(tag, env.body);
  for (const p of env.proofs) {
    const pk = keyBytes(p.kid);
    const sig = b64uDecode(p.signature);
    if (pk === null || sig === null || sig.length !== 64) return "SIGNATURE_INVALID";
    if (!verifyFn(pk, digest, sig)) return "SIGNATURE_INVALID";
  }
  return null;
}

/** Produce one proof for a body under one seed. */
export function signBody(tag: Tag, body: JsonObject, kid: string, seed: Uint8Array): { kid: string; signature: string } {
  const d = digestBytes(tag, body);
  const sig = edSign(seed, d);
  let bin = "";
  for (const b of sig) bin += String.fromCharCode(b);
  return { kid, signature: btoa(bin).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "") };
}

/** Build a Signed<T> envelope from a body and a set of (kid, seed) signers. */
export function signed(tag: Tag, body: JsonObject, signers: { kid: string; seed: Uint8Array }[]): SignedEnvelope {
  const proofs = signers
    .map((s) => signBody(tag, body, s.kid, s.seed))
    .sort((a, b) => (a.kid < b.kid ? -1 : a.kid > b.kid ? 1 : 0));
  const seen = new Set<string>();
  const out = proofs.filter((p) => (seen.has(p.kid) ? false : (seen.add(p.kid), true)));
  return { body: JSON.parse(JSON.stringify(body)) as JsonObject, proofs: out };
}

export function signedBodyHash(tag: Tag, env: SignedEnvelope): string {
  return digest(tag, env.body);
}

File: python/src/latticeag_herald/proofs.py
"""Proof-set verification (spec §3, §5): proofs are sorted ascending by kid,
unique, and must equal exactly the required signer set. Each proof signs
the 32 raw bytes of D(tag, body).
"""

import copy

from .b64u import b64u_decode, b64u_encode
from .digest import digest, digest_bytes
from .ed25519 import ed25519_sign, ed25519_verify


def verify_proof_set(tag, env, expected_kids, key_bytes):
    """Check the proof set equals expected_kids (compared as a set after dedup
    of expected values), then verify every signature.
    Returns None on success, or "PROOF_SET_INVALID" / "SIGNATURE_INVALID"."""
    expected = sorted(set(expected_kids))
    actual = [p["kid"] for p in env["proofs"]]
    if actual != sorted(actual) or len(set(actual)) != len(actual):
        return "PROOF_SET_INVALID"
    if actual != expected:
        return "PROOF_SET_INVALID"
    d = digest_bytes(tag, env["body"])
    for p in env["proofs"]:
        pk = key_bytes(p["kid"])
        sig = b64u_decode(p["signature"])
        if pk is None or sig is None or len(sig) != 64:
            return "SIGNATURE_INVALID"
        if not ed25519_verify(pk, d, sig):
            return "SIGNATURE_INVALID"
    return None


def sign_body(tag, body, kid, seed):
    d = digest_bytes(tag, body)
    return {"kid": kid, "signature": b64u_encode(ed25519_sign(seed, d))}


def signed(tag, body, signers):
    """Build a signed envelope from a body and a list of (kid, seed) pairs."""
    seen = set()
    proofs = []
    for kid, seed in sorted(signers, key=lambda s: s[0]):
        if kid in seen:
            continue
        seen.add(kid)
        proofs.append(sign_body(tag, body, kid, seed))
    return {"body": copy.deepcopy(body), "proofs": proofs}


def signed_body_hash(tag, env):
    return digest(tag, env["body"])
